contracts/truefi2/CrvBaseRateOracle.sol

```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.6.10;
+
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
+import "../truefi/interface/ICurve.sol";
+
+// prettier-ignore
+contract CrvBaseRateOracle {
+    using SafeMath for uint256;
+    uint8 public constant BUFFER_SIZE = 7;
+
+    ICurve public curve;
+
+    struct HistoricalRatesBuffer {
+        uint256[BUFFER_SIZE] baseRates;
+        uint256[BUFFER_SIZE] timestamps;
+        uint8 insertIndex;
+    }
+
+    HistoricalRatesBuffer public histBuffer;
+
+    constructor(ICurve _curve) public {
+        curve = _curve;
+    }
+
+    function getHistBuffer() public view returns (uint256[BUFFER_SIZE] memory, uint256[BUFFER_SIZE] memory, uint8) {
+        return (histBuffer.baseRates, histBuffer.timestamps, histBuffer.insertIndex);
+    }
+
+    function updateRate() public {
+        uint8 iidx = histBuffer.insertIndex;
+        histBuffer.timestamps[iidx] = block.timestamp;
+        histBuffer.baseRates[iidx] = curve.get_virtual_price();
+        histBuffer.insertIndex = (iidx + 1) % BUFFER_SIZE;
+    }
+
+    function calculateAverageRate() public view returns (uint256) {
+        // Average rate is calculated based on formula below:
+        //           v_n - v_1
+        // avgRate = ---------
+        //           t_n - t_1
+        //
+        // where v_n is the most recent rate, v_1 is the oldest rate,
+        // t_n and t_1 are their timestamps respectively
+        uint8 idxOfOldestRate = histBuffer.insertIndex;
+        uint256 v_n = histBuffer.baseRates[(idxOfOldestRate - 1) % BUFFER_SIZE];
```

Comment:
This will result in underflow if idxOfOldestRate is 0. And 255 % 7 = 3 instead of expected 7 so we need to make additional checks for this case